File: algorithms/fixed_window_counter.lua
-- KEYS[1] = redis key (includes window id)
-- ARGV[1] = limit
-- ARGV[2] = window_size (seconds)
-- ARGV[3] = now (epoch seconds)

local limit = tonumber(ARGV[1])
local window_size = tonumber(ARGV[2])
local now = tonumber(ARGV[3])

-- Increment counter
local current = redis.call("INCR", KEYS[1])

-- Set expiry on first request
if current == 1 then
    redis.call("EXPIRE", KEYS[1], window_size)
end

if current <= limit then
    return {1, current, 0}  -- allowed, retry_after = 0
else
    -- Compute window end
    local window_start = now - (now % window_size)
    local window_end = window_start + window_size
    local retry_after = window_end - now

    return {0, current, retry_after}
end

File: algorithms/fixed_window_counter.py
from time import time
from redis_client import run_fixed_window_counter

def fixed_window_counter_limit(
    *,
    identifier: str,
    capacity: int,
    window_size: int,
) -> tuple[bool, dict]:
    """
    Apply fixed window counter rate limiting.

    identifier   -> user/IP/route key
    capacity     -> max requests allowed in the window
    window_size  -> size of the time window in seconds
    """
    now = int(time())
    redis_key = f"rate:fixed_window_counter:{identifier}"

    allowed,curr_capacity,retry_after = run_fixed_window_counter(
        key=redis_key,
        capacity=capacity,
        window_size=window_size,
        now=now,
    )

    return allowed, {"current": curr_capacity, "retry_after": retry_after}